nexus/pkg/artifacts/downloader.go

```diff
@@ -0,0 +1,309 @@
+	manifest, err := loadManifestFromURL(directURL)
+	if err != nil {
+		return Manifest{}, err
+	}
+
+	// Set upstream artifacts, if any
+	err = ad.setUpstreamArtifacts(manifest)
+	if err != nil {
+		return Manifest{}, err
+	}
+	return manifest, nil
+}
+
+func (ad *ArtifactDownloader) setDefaultDownloadRoot(artifactAttrs gql.ArtifactByNameProjectArtifact) (rerr error) {
+	// This sets the absolute path of source_artifact_name:v{versionIndex} as default download root
+	sourceArtifactName := artifactAttrs.GetArtifactSequence().Name
+	versionIndex := artifactAttrs.GetVersionIndex()
+	if versionIndex == nil {
+		return fmt.Errorf("invalid artifact download, missing version index")
+	}
+	artifactDir, err := env.GetArtifactDir()
+	if err != nil {
+		return err
+	}
+	downloadRoot := filepath.Join(artifactDir, fmt.Sprintf("%s:v%d", sourceArtifactName, *versionIndex))
+
+	path := CheckExists(downloadRoot)
+	if path == nil {
+		downloadRoot = SystemPreferredPath(downloadRoot, false)
+	}
+	ad.DefaultDownloadRoot = &downloadRoot
+	return nil
+}
+
+func (ad *ArtifactDownloader) setUpstreamArtifacts(manifest Manifest) error {
+	for _, entry := range manifest.Contents {
+		referencedID, err := getReferencedID(entry.Ref)
+		if err != nil {
+			return err
+		}
+		if referencedID != nil {
+			response, err := gql.ArtifactByID(
+				ad.Ctx,
+				ad.GraphqlClient,
+				*referencedID,
+			)
+			if err != nil {
+				return err
+			} else if response == nil {
+				return fmt.Errorf("could not get artifact by id for reference %s", *entry.Ref)
+			}
+			depArtifact := response.GetArtifact()
+			if depArtifact != nil {
+				ad.UpstreamArtifacts = append(ad.UpstreamArtifacts, *depArtifact)
+			}
+		}
+	}
+	return nil
+}
+
+func (ad *ArtifactDownloader) downloadFiles(artifactID string, manifest Manifest) error {
+	var downloadRoot string
+	if ad.DownloadRoot != nil {
+		downloadRoot = *ad.DownloadRoot
+	} else if ad.DefaultDownloadRoot != nil {
+		downloadRoot = *ad.DefaultDownloadRoot
+	}
+	if downloadRoot == "" {
+		return fmt.Errorf("download root was not set")
+	}
+	// retrieve from "WANDB_ARTIFACT_FETCH_FILE_URL_BATCH_SIZE"?
+	batchSize := BATCH_SIZE
+
+	type TaskResult struct {
+		Task *filetransfer.Task
+		Name string
+	}
+
+	// Fetch URLs and download files in batches
+	manifestEntries := manifest.Contents
+	numInProgress, numDone := 0, 0
+	nameToScheduledTime := map[string]time.Time{}
+	taskResultsChan := make(chan TaskResult)
+	manifestEntriesBatch := make([]ManifestEntry, 0, batchSize)
+
+	for numDone < len(manifestEntries) {
+		cursor := ""
+		hasNextPage := true
+		for hasNextPage {
+			// Prepare a batch
+			now := time.Now()
+			manifestEntriesBatch = manifestEntriesBatch[:0]
+			response, err := gql.ArtifactFileURLs(
+				ad.Ctx,
+				ad.GraphqlClient,
+				artifactID,
+				&cursor,
+				&batchSize,
+			)
+			if err != nil {
+				return err
+			} else if response == nil {
```

Comment:
Nit: `response` is never `nil`, you don't have to check it. Look at the implementation of `ArtifactFileURLs`. Same for all other GraphQL queries.